## Arturo Ramos Viedas
# A01636133

# Data management using Pandas

**Data management** is a crucial component to statistical analysis and data science work.

This notebook will show you how to import, view, undertand, and manage your data using the [Pandas](http://pandas.pydata.org) data processing library, i.e., the notebook will demonstrates how to read a dataset into Python, and obtain a basic understanding of its content.

Note that **Python** by itself is a general-purpose programming language and does not provide high-level data processing capabilities.  The **Pandas** library was developed to meet this need. **Pandas** is the most popular Python library for data manipulation, and we will use it extensively in this course. **Pandas** provides high-performance, easy-to-use data structures and data analysis tools.

The main data structure that **Pandas** works with is called a **Data Frame**. This is a two-dimensional table of data in which the rows typically
represent cases and the columns represent variables (e.g. data used in this tutorial).  Pandas also has a one-dimensional data structure called a **Series** that we will encounter when accesing a single column of a Data Frame.

Pandas has a variety of functions named `read_xxx` for reading data in different formats.  Right now we will focus on reading `csv` files, which stands for comma-separated values. However the other file formats include `excel`, `json`, and `sql`.

There are many other options to `read_csv` that are very useful.  For example, you would use the option `sep='\t'` instead of the default `sep=','` if the fields of your data file are delimited by tabs instead of commas.  See [here](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.read_csv.html) for the full documentation for `read_csv`.


## Acknowledgments

- The dataset used in this tutorial is from https://www.coursera.org/ from the course "Understanding and Visualizing Data with Python" by University of Michigan


# Importing libraries


In [3]:
# Import the packages that we will be using
import pandas as pd                 # For data handling

#pd.set_option('display.max_columns', 100) # Show all columns when looking at dataframe

# Importing data

In [7]:
# Define where you are running the code: colab or local
RunInColab          = True     # (False: no  | True: yes)

# If running in colab:
if RunInColab:
    # Mount your google drive in google colab
    from google.colab import drive
    drive.mount('/content/drive')

    # Find location
    #!pwd
    #!ls
    #!ls "/content/drive/My Drive/Colab Notebooks/MachineLearningWithPython/"

    # Define path del proyecto
    Ruta            = "/content/drive/My Drive/A01636133/"

else:
    # Define path del proyecto
    Ruta            = ""

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [83]:
# url string that hosts our .csv file
url = Ruta + "datasets/iris/iris.csv"
colnames = ['SepalLength', 'SepalWidth', 'PetalLength', 'PetalWidth', 'Flower']

# Read the .csv file and store it as a pandas Data Frame
df = pd.read_csv(url, names=colnames, header=None)



If we want to print the information about th output object type we would simply type the following:

In [47]:
type(df)

pandas.core.frame.DataFrame

# Exploring the content of the data set

Use the `shape` method to determine the numbers of rows and columns in a data frame. This can be used to confirm that we have actually obtained the data the we are expecting.

Based on what we see below, the data set being read here has $N_r$ rows, corresponding to $N_r$ observations, and $N_c$ columns, corresponding to $N_c$ variables in this particular data file.

In [48]:
df.shape

(150, 5)

In [49]:
# Print the number of rows
Nrows = df.shape[0]
Nrows

150

In [50]:
# Print the number of columns
Ncols = df.shape[1]
Ncols

5

If we want to show the entire data frame we would simply write the following:

In [51]:
df

SepalLength  SepalWidth  PetalLength  PetalWidth           Class
0            5.1         3.5          1.4         0.2     Iris-setosa
1            4.9         3.0          1.4         0.2     Iris-setosa
2            4.7         3.2          1.3         0.2     Iris-setosa
3            4.6         3.1          1.5         0.2     Iris-setosa
4            5.0         3.6          1.4         0.2     Iris-setosa
..           ...         ...          ...         ...             ...
145          6.7         3.0          5.2         2.3  Iris-virginica
146          6.3         2.5          5.0         1.9  Iris-virginica
147          6.5         3.0          5.2         2.0  Iris-virginica
148          6.2         3.4          5.4         2.3  Iris-virginica
149          5.9         3.0          5.1         1.8  Iris-virginica

[150 rows x 5 columns]

As you can see, we have a 2-Dimensional object where each row is an independent observation and each coloum is a variable.

Now, use the the `head()` function to show the first 5 rows of our data frame

In [52]:
df.head()
#df.head(10)

SepalLength  SepalWidth  PetalLength  PetalWidth        Class
0          5.1         3.5          1.4         0.2  Iris-setosa
1          4.9         3.0          1.4         0.2  Iris-setosa
2          4.7         3.2          1.3         0.2  Iris-setosa
3          4.6         3.1          1.5         0.2  Iris-setosa
4          5.0         3.6          1.4         0.2  Iris-setosa

Also, you can use the the `tail()` function to show the last 5 rows of our data frame

In [53]:
df.tail()
#df.tail(3)

SepalLength  SepalWidth  PetalLength  PetalWidth           Class
145          6.7         3.0          5.2         2.3  Iris-virginica
146          6.3         2.5          5.0         1.9  Iris-virginica
147          6.5         3.0          5.2         2.0  Iris-virginica
148          6.2         3.4          5.4         2.3  Iris-virginica
149          5.9         3.0          5.1         1.8  Iris-virginica

The columns in a Pandas data frame have names, to see the names, use the `columns` method:

To gather more information regarding the data, we can view the column names with the following function:

In [54]:
df.columns

Index(['SepalLength', 'SepalWidth', 'PetalLength', 'PetalWidth', 'Class'], dtype='object')

Be aware that every variable in a Pandas data frame has a data type.  There are many different data types, but most commonly you will encounter floating point values (real numbers), integers, strings (text), and date/time values.  When Pandas reads a text/csv file, it guesses the data types based on what it sees in the first few rows of the data file.  Usually it selects an appropriate type, but occasionally it does not.  To confirm that the data types are consistent with what the variables represent, inspect the `dtypes` attribute of the data frame.

In [55]:
df.dtypes

SepalLength    float64
SepalWidth     float64
PetalLength    float64
PetalWidth     float64
Class           object
dtype: object

Summary statistics, which include things like the mean, min, and max of the data, can be useful to get a feel for how large some of the variables are and what variables may be the most important.

In [65]:
# Summary statistics for the quantitative variables
df.describe()


SepalLength  SepalWidth  PetalLength  PetalWidth
count   150.000000  150.000000   150.000000  150.000000
mean      5.843333    3.057333     3.758000    1.199333
std       0.828066    0.435866     1.765298    0.762238
min       4.300000    2.000000     1.000000    0.100000
25%       5.100000    2.800000     1.600000    0.300000
50%       5.800000    3.000000     4.350000    1.300000
75%       6.400000    3.300000     5.100000    1.800000
max       7.900000    4.400000     6.900000    2.500000

In [57]:
# Drop observations with NaN values
df.SepalLength.dropna().describe()
#df.Wingspan.dropna().describe()

count    150.000000
mean       5.843333
std        0.828066
min        4.300000
25%        5.100000
50%        5.800000
75%        6.400000
max        7.900000
Name: SepalLength, dtype: float64

It is also possible to get statistics on the entire data frame or a column as follows

- `df.mean()` Returns the mean of all columns
- `df.corr()` Returns the correlation between columns in a data frame
- `df.count()` Returns the number of non-null values in each data frame column
- `df.max()` Returns the highest value in each column
- `df.min()` Returns the lowest value in each column
- `df.median()` Returns the median of each column
- `df.std()` Returns the standard deviation of each column

In [58]:
df.mean()

<ipython-input-58-c61f0c8f89b5>:1: FutureWarning: The default value of numeric_only in DataFrame.mean is deprecated. In a future version, it will default to False. In addition, specifying 'numeric_only=None' is deprecated. Select only valid columns or specify the value of numeric_only to silence this warning.
  df.mean()


SepalLength    5.843333
SepalWidth     3.057333
PetalLength    3.758000
PetalWidth     1.199333
dtype: float64

# How to write a data frame to a File

To save a file with your data simply use the `to_csv` attribute

In [ ]:
df.to_csv('myDataFrame.csv')
#df.to_csv('myDataFrame.csv', sep='\t')

# Rename columns

To change the name of a colum use the `rename` attribute

In [ ]:
df = df.rename(columns={"Age": "Edad"})

df.head()

ID  Edad Gender  GenderGroup Glasses  GlassesGroup  Height  Wingspan  \
0   1  56.0      F            1       Y             1    62.0      61.0   
1   2  26.0      F            1       Y             1    62.0      60.0   
2   3  33.0      F            1       Y             1    66.0      64.0   
3   4  39.0      F            1       N             0    64.0      63.0   
4   5  27.0      M            2       N             0    73.0      75.0   

   CWDistance Complete  CompleteGroup  Score  
0          79        Y            1.0      7  
1          70        Y            1.0      8  
2          85        Y            1.0      7  
3          87        Y            1.0     10  
4          72        N            0.0      4

In [ ]:
df = df.rename(columns={"Edad": "Age"})

df.head()

ID   Age Gender  GenderGroup Glasses  GlassesGroup  Height  Wingspan  \
0   1  56.0      F            1       Y             1    62.0      61.0   
1   2  26.0      F            1       Y             1    62.0      60.0   
2   3  33.0      F            1       Y             1    66.0      64.0   
3   4  39.0      F            1       N             0    64.0      63.0   
4   5  27.0      M            2       N             0    73.0      75.0   

   CWDistance Complete  CompleteGroup  Score  
0          79        Y            1.0      7  
1          70        Y            1.0      8  
2          85        Y            1.0      7  
3          87        Y            1.0     10  
4          72        N            0.0      4

# Selection of colums

As discussed above, a Pandas data frame is a rectangular data table, in which the rows represent observations or samples and the columns represent variables.  One common manipulation of a data frame is to extract the data for one case or for one variable.  There are several ways to do this, as shown below.

To extract all the values for one column (variable), use one of the following alternatives.

In [ ]:
a = df.Age
b = df["Age"]
c = df.loc[:, "Age"]
d = df.iloc[:, 1]

print(d)

#df[["Gender", "GenderGroup"]]

0     56.0
1     26.0
2     33.0
3     39.0
4     27.0
5     24.0
6     28.0
7     22.0
8     29.0
9     33.0
10    30.0
11    28.0
12    25.0
13    23.0
14    31.0
15    26.0
16    26.0
17    27.0
18    23.0
19    24.0
20    23.0
21    29.0
22    25.0
23    26.0
24    23.0
25    28.0
26    24.0
27    25.0
28    32.0
29    38.0
30    27.0
31    33.0
32    38.0
33    27.0
34    24.0
35    27.0
36    25.0
37    26.0
38    31.0
39    30.0
40    23.0
41    26.0
42    28.0
43    26.0
44    30.0
45    39.0
46    27.0
47    24.0
48    28.0
49    30.0
50     NaN
51    27.0
Name: Age, dtype: float64


# Slicing a data set

As discussed above, a Pandas data frame is a rectangular data table, in which the rows represent cases and the columns represent variables.  One common manipulation of a data frame is to extract the data for one observation or for one variable.  There are several ways to do this, as shown below.

Lets say we would like to splice our data frame and select only specific portions of our data.  There are three different ways of doing so.

1. .loc()
2. .iloc()
3. .ix()

We will cover the .loc() and .iloc() splicing functions.


The attibute **.loc()** uses labels/column names, in specific, it takes two single/list/range operator separated by ',', the first one indicates the rows and the second one indicates columns.

In [ ]:
# Return all observations of CWDistance
df.loc[:,"CWDistance"]

# Return a subset of observations of CWDistance
df.loc[:9, "CWDistance"]

# Select all rows for multiple columns, ["Gender", "GenderGroup"]
df.loc[:,["Gender", "GenderGroup"]]

# Select multiple columns, ["Gender", "GenderGroup"]me
keep = ['Gender', 'GenderGroup']
df_gender = df[keep]

# Select few rows for multiple columns, ["CWDistance", "Height", "Wingspan"]
#df.loc[4:9, ["CWDistance", "Height", "Wingspan"]]

# Select range of rows for all columns
#df.loc[10:15,:]



The attribute **iloc()** is an integer based slicing.

In [ ]:
# .
df.iloc[:, :4]

# .
df.iloc[:4, :]

# .
df.iloc[:, 3:7]

# .
df.iloc[4:8, 2:4]

# This is incorrect:
#df.iloc[1:5, ["Gender", "GenderGroup"]]

Gender  GenderGroup
4      M            2
5      M            2
6      M            2
7      F            1

# Get unique existing values

List unique values in the one of the columns


In [ ]:
# List unique values in the df['Gender'] column
df.Gender.unique()

array(['F', 'M'], dtype=object)

In [ ]:
# Lets explore df["GenderGroup] as well
df.GenderGroup.unique()

array([1, 2])

# Filter, Sort and Groupby



With **Filter** you can use different conditions to filter columns. For example, df[df[year] > 1984] would give you only the column year is greater than 1984. You can use & (and) or | (or) to add different conditions to your filtering. This is also called boolean filtering.

In [ ]:
df[df["Height"] >= 70]

ID   Age Gender  GenderGroup Glasses  GlassesGroup  Height  Wingspan  \
4    5  27.0      M            2       N             0    73.0      75.0   
5    6  24.0      M            2       N             0    75.0      71.0   
6    7  28.0      M            2       N             0    75.0      76.0   
8    9  29.0      M            2       Y             1    74.0      73.0   
14  15  31.0      M            2       Y             1    73.0      74.0   
15  16  26.0      M            2       Y             1    71.0      72.0   
18  19  23.0      M            2       Y             1    70.0      69.0   
21  22  29.0      M            2       N             0    71.0      70.0   
22  23  25.0      M            2       N             0    70.0      68.0   
25  26  28.0      M            2       N             0    75.0      76.0   
26  27  24.0      M            2       N             0    78.4      71.0   
27  28  25.0      M            2       Y             1    76.0      73.0   
33  34  27.0      M            2       N             0    77.0      75.0   
38  39  31.0      M            2       Y             1    73.0      74.0   
40  41  23.0      F            1       N             0    70.4      71.0   
41  42  26.0      M            2       Y             1    73.5      72.0   
42  43  28.0      F            1       Y             1    72.5      72.0   
43  44  26.0      F            1       Y             1    72.0      72.0   
46  47  27.0      M            2       N             0    78.0      75.0   
47  48  24.0      M            2       N             0    79.5      75.0   
48  49  28.0      M            2       N             0    77.8      76.0   
49  50  30.0      F            1       N             0    74.6       NaN   
50  51   NaN      M            2       N             0    71.0      70.0   

    CWDistance Complete  CompleteGroup  Score  
4           72        N            0.0      4  
5           81        N            0.0      3  
6          107        Y            1.0     10  
8          106        N            0.0      5  
14          72        Y            1.0      9  
15         115        Y            1.0      6  
18          64        Y            1.0      3  
21         101        Y            1.0      8  
22          82        Y            1.0      4  
25         111        Y            1.0     10  
26          92        Y            1.0      7  
27         107        Y            1.0      8  
33         100        Y            1.0      8  
38          72        Y            1.0      9  
40          66        Y            1.0      4  
41         115        Y            1.0      6  
42          81        Y            1.0     10  
43          92        Y            1.0      8  
46          72        N            0.0      7  
47          82        N            0.0      8  
48          99        Y            1.0      9  
49          71        Y            1.0      9  
50         101        Y            NaN      8

With **Sort** is possible to sort values in a certain column in an ascending order using `df.sort_values("ColumnName")` or in descending order using `df.sort_values(ColumnName, ascending=False)`.

Furthermore, it’s possible to sort values by Column1Name in ascending order then Column2Name in descending order by using `df.sort_values([Column1Name,Column2Name],ascending=[True,False])`


In [ ]:
df.sort_values("Height")
#df.sort_values("Height",ascending=False)

ID   Age Gender  GenderGroup Glasses  GlassesGroup  Height  Wingspan  \
16  17  26.0      F            1       N             0   61.50      59.5   
29  30  38.0      F            1       Y             1   61.50      61.0   
37  38  26.0      F            1       N             0   61.50      59.5   
13  14  23.0      F            1       N             0   61.50      57.5   
0    1  56.0      F            1       Y             1   62.00      61.0   
30  31  27.0      F            1       Y             1   62.00      60.0   
1    2  26.0      F            1       Y             1   62.00      60.0   
11  12  28.0      F            1       Y             1   62.75      58.0   
28  29  32.0      F            1       Y             1   63.00      60.0   
9   10  33.0      F            1       Y             1   63.00      60.0   
32  33  38.0      F            1       N             0   64.00      63.0   
3    4  39.0      F            1       N             0   64.00      63.0   
45  46  39.0      F            1       N             0   64.00      63.0   
24  25  23.0      F            1       Y             1   65.00      63.0   
36  37  25.0      F            1       Y             1   65.00      64.5   
7    8  22.0      F            1       N             0   65.00      62.0   
12  13  25.0      F            1       Y             1   65.00      64.5   
31  32  33.0      F            1       Y             1   65.30      64.0   
2    3  33.0      F            1       Y             1   66.00      64.0   
17  18  27.0      M            2       N             0   66.00      66.0   
44  45  30.0      F            1       Y             1   66.00      64.0   
34  35  24.0      F            1       N             0   67.80      62.0   
19  20  24.0      F            1       Y             1   68.00      66.0   
35  36  27.0      M            2       N             0   68.00      66.0   
20  21  23.0      M            2       Y             1   69.00      67.0   
23  24  26.0      M            2       N             0   69.00      71.0   
10  11  30.0      M            2       Y             1   69.50      66.0   
39  40  30.0      M            2       Y             1   69.50      66.0   
18  19  23.0      M            2       Y             1   70.00      69.0   
22  23  25.0      M            2       N             0   70.00      68.0   
40  41  23.0      F            1       N             0   70.40      71.0   
50  51   NaN      M            2       N             0   71.00      70.0   
21  22  29.0      M            2       N             0   71.00      70.0   
15  16  26.0      M            2       Y             1   71.00      72.0   
43  44  26.0      F            1       Y             1   72.00      72.0   
42  43  28.0      F            1       Y             1   72.50      72.0   
4    5  27.0      M            2       N             0   73.00      75.0   
14  15  31.0      M            2       Y             1   73.00      74.0   
38  39  31.0      M            2       Y             1   73.00      74.0   
41  42  26.0      M            2       Y             1   73.50      72.0   
8    9  29.0      M            2       Y             1   74.00      73.0   
49  50  30.0      F            1       N             0   74.60       NaN   
5    6  24.0      M            2       N             0   75.00      71.0   
25  26  28.0      M            2       N             0   75.00      76.0   
6    7  28.0      M            2       N             0   75.00      76.0   
27  28  25.0      M            2       Y             1   76.00      73.0   
33  34  27.0      M            2       N             0   77.00      75.0   
48  49  28.0      M            2       N             0   77.80      76.0   
46  47  27.0      M            2       N             0   78.00      75.0   
26  27  24.0      M            2       N             0   78.40      71.0   
47  48  24.0      M            2       N             0   79.50      75.0   
51  52  27.0      M            2       N            

The attribute **Groupby** involves splitting the data into groups based on some criteria, applying a function to each group independently and combining the results into a data structure. df.groupby(col) returns a groupby object for values from one column while df.groupby([col1,col2]) returns a groupby object for values from multiple columns.

In [ ]:
df.groupby(['Gender'])


Size of each group

In [ ]:
df.groupby(['Gender']).size()

df.groupby(['Gender','GenderGroup']).size()

Gender  GenderGroup
F       1              26
M       2              26
dtype: int64

This output indicates that we have two types of combinations.

* Case 1: Gender = F & Gender Group = 1
* Case 2: Gender = M & GenderGroup = 2.  

This validates our initial assumption that these two fields essentially portray the same information.

# Data Cleaning: handle with missing data

Before getting started to work with your data, it's a good practice to observe it thoroughly to identify missing values and handle them accordingly.

When reading a dataset using Pandas, there is a set of values including 'NA', 'NULL', and 'NaN' that are taken by default to represent a missing value.  The full list of default missing value codes is in the '`read_csv`' documentation [here](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.read_csv.html).  This document also explains how to change the way that '`read_csv`' decides whether a variable's value is missing.

Pandas has functions called `isnull` and `notnull` that can be used to identify where the missing and non-missing values are located in a data frame.  

Below we use these functions to count the number of missing and non-missing values in each variable of the datasetr.

In [59]:
df.isnull()


SepalLength  SepalWidth  PetalLength  PetalWidth  Class
0          False       False        False       False  False
1          False       False        False       False  False
2          False       False        False       False  False
3          False       False        False       False  False
4          False       False        False       False  False
..           ...         ...          ...         ...    ...
145        False       False        False       False  False
146        False       False        False       False  False
147        False       False        False       False  False
148        False       False        False       False  False
149        False       False        False       False  False

[150 rows x 5 columns]

Unfortunately, our output indicates that some of our columns contain missing values so we are no able to continue on doing analysis with those colums

In [60]:
df.notnull()

SepalLength  SepalWidth  PetalLength  PetalWidth  Class
0           True        True         True        True   True
1           True        True         True        True   True
2           True        True         True        True   True
3           True        True         True        True   True
4           True        True         True        True   True
..           ...         ...          ...         ...    ...
145         True        True         True        True   True
146         True        True         True        True   True
147         True        True         True        True   True
148         True        True         True        True   True
149         True        True         True        True   True

[150 rows x 5 columns]

In [61]:
df.isnull().sum()
df.notnull().sum()

SepalLength    150
SepalWidth     150
PetalLength    150
PetalWidth     150
Class          150
dtype: int64

Now we use these functions to count the number of missing and non-missing values in a single variable in the dataset

In [63]:
print( df.SepalLength.notnull().sum() )

print( pd.isnull(df.SepalLength).sum() )

150
0


In [ ]:
# Extract all non-missing values of one of the columns into a new variable
x = df.Age.dropna().describe()
x.describe()

count     8.000000
mean     30.645922
std      16.044470
min       5.755611
25%      24.250000
50%      27.705882
75%      35.250000
max      56.000000
Name: Age, dtype: float64

# Add and eliminate columns

In some cases it is useful to create or eiminate new columns

In [64]:
df.head()

SepalLength  SepalWidth  PetalLength  PetalWidth        Class
0          5.1         3.5          1.4         0.2  Iris-setosa
1          4.9         3.0          1.4         0.2  Iris-setosa
2          4.7         3.2          1.3         0.2  Iris-setosa
3          4.6         3.1          1.5         0.2  Iris-setosa
4          5.0         3.6          1.4         0.2  Iris-setosa

In [ ]:
# Add a new column with new data

# Create a column data
NewColumnData = df.Age/df.Age

# Insert that column in the data frame
df.insert(12, "ColumnInserted", NewColumnData, True)

df.head()

ID   Age Gender  GenderGroup Glasses  GlassesGroup  Height  Wingspan  \
0   1  56.0      F            1       Y             1    62.0      61.0   
1   2  26.0      F            1       Y             1    62.0      60.0   
2   3  33.0      F            1       Y             1    66.0      64.0   
3   4  39.0      F            1       N             0    64.0      63.0   
4   5  27.0      M            2       N             0    73.0      75.0   

   CWDistance Complete  CompleteGroup  Score  ColumnInserted  
0          79        Y            1.0      7             1.0  
1          70        Y            1.0      8             1.0  
2          85        Y            1.0      7             1.0  
3          87        Y            1.0     10             1.0  
4          72        N            0.0      4             1.0

In [ ]:
# Eliminate inserted column
df.drop("ColumnInserted", axis=1, inplace = True)
#df.drop(columns=['ColumnInserted'], inplace = True)
# Remove three columns as index base
#df.drop(df.columns[[12]], axis = 1, inplace = True)

df.head()

ID   Age Gender  GenderGroup Glasses  GlassesGroup  Height  Wingspan  \
0   1  56.0      F            1       Y             1    62.0      61.0   
1   2  26.0      F            1       Y             1    62.0      60.0   
2   3  33.0      F            1       Y             1    66.0      64.0   
3   4  39.0      F            1       N             0    64.0      63.0   
4   5  27.0      M            2       N             0    73.0      75.0   

   CWDistance Complete  CompleteGroup  Score  
0          79        Y            1.0      7  
1          70        Y            1.0      8  
2          85        Y            1.0      7  
3          87        Y            1.0     10  
4          72        N            0.0      4

In [ ]:
# Add new column derived from existing columns

# The new column is a function of another column
df["AgeInMonths"] = df["Age"] * 12

df.head()

ID   Age Gender  GenderGroup Glasses  GlassesGroup  Height  Wingspan  \
0   1  56.0      F            1       Y             1    62.0      61.0   
1   2  26.0      F            1       Y             1    62.0      60.0   
2   3  33.0      F            1       Y             1    66.0      64.0   
3   4  39.0      F            1       N             0    64.0      63.0   
4   5  27.0      M            2       N             0    73.0      75.0   

   CWDistance Complete  CompleteGroup  Score  AgeInMonths  
0          79        Y            1.0      7        672.0  
1          70        Y            1.0      8        312.0  
2          85        Y            1.0      7        396.0  
3          87        Y            1.0     10        468.0  
4          72        N            0.0      4        324.0

In [ ]:
# Eliminate inserted column
df.drop("AgeInMonths", axis=1, inplace = True)

df.head()

ID   Age Gender  GenderGroup Glasses  GlassesGroup  Height  Wingspan  \
0   1  56.0      F            1       Y             1    62.0      61.0   
1   2  26.0      F            1       Y             1    62.0      60.0   
2   3  33.0      F            1       Y             1    66.0      64.0   
3   4  39.0      F            1       N             0    64.0      63.0   
4   5  27.0      M            2       N             0    73.0      75.0   

   CWDistance Complete  CompleteGroup  Score  
0          79        Y            1.0      7  
1          70        Y            1.0      8  
2          85        Y            1.0      7  
3          87        Y            1.0     10  
4          72        N            0.0      4

In [ ]:
# Add a new column with text labels reflecting the code's meaning

df["GenderGroupNew"] = df.GenderGroup.replace({1: "Female", 2: "Male"})

# Show the first 5 rows of the created data frame
df.head()

ID   Age Gender  GenderGroup Glasses  GlassesGroup  Height  Wingspan  \
0   1  56.0      F            1       Y             1    62.0      61.0   
1   2  26.0      F            1       Y             1    62.0      60.0   
2   3  33.0      F            1       Y             1    66.0      64.0   
3   4  39.0      F            1       N             0    64.0      63.0   
4   5  27.0      M            2       N             0    73.0      75.0   

   CWDistance Complete  CompleteGroup  Score GenderGroupNew  
0          79        Y            1.0      7         Female  
1          70        Y            1.0      8         Female  
2          85        Y            1.0      7         Female  
3          87        Y            1.0     10         Female  
4          72        N            0.0      4           Male

In [ ]:
# Eliminate inserted column
df.drop("GenderGroupNew", axis=1, inplace = True)
#df.drop(['GenderGroupNew'],vaxis='columns',vinplace=True)

df.head()

ID   Age Gender  GenderGroup Glasses  GlassesGroup  Height  Wingspan  \
0   1  56.0      F            1       Y             1    62.0      61.0   
1   2  26.0      F            1       Y             1    62.0      60.0   
2   3  33.0      F            1       Y             1    66.0      64.0   
3   4  39.0      F            1       N             0    64.0      63.0   
4   5  27.0      M            2       N             0    73.0      75.0   

   CWDistance Complete  CompleteGroup  Score  
0          79        Y            1.0      7  
1          70        Y            1.0      8  
2          85        Y            1.0      7  
3          87        Y            1.0     10  
4          72        N            0.0      4

In [ ]:
# Add a new column with strata based on these cut points

# Create a column data
NewColumnData = df.Age/df.Age

# Insert that column in the data frame
df.insert(1, "ColumnStrata", NewColumnData, True)

df["ColumnStrata"] = pd.cut(df.Height, [60., 63., 66., 69., 72., 75., 78.])

# Show the first 5 rows of the created data frame
df.head()

ID  ColumnStrata   Age Gender  GenderGroup Glasses  GlassesGroup  Height  \
0   1  (60.0, 63.0]  56.0      F            1       Y             1    62.0   
1   2  (60.0, 63.0]  26.0      F            1       Y             1    62.0   
2   3  (63.0, 66.0]  33.0      F            1       Y             1    66.0   
3   4  (63.0, 66.0]  39.0      F            1       N             0    64.0   
4   5  (72.0, 75.0]  27.0      M            2       N             0    73.0   

   Wingspan  CWDistance Complete  CompleteGroup  Score  
0      61.0          79        Y            1.0      7  
1      60.0          70        Y            1.0      8  
2      64.0          85        Y            1.0      7  
3      63.0          87        Y            1.0     10  
4      75.0          72        N            0.0      4

In [ ]:
# Eliminate inserted column
df.drop("ColumnStrata", axis=1, inplace = True)

df.head()

ID   Age Gender  GenderGroup Glasses  GlassesGroup  Height  Wingspan  \
0   1  56.0      F            1       Y             1    62.0      61.0   
1   2  26.0      F            1       Y             1    62.0      60.0   
2   3  33.0      F            1       Y             1    66.0      64.0   
3   4  39.0      F            1       N             0    64.0      63.0   
4   5  27.0      M            2       N             0    73.0      75.0   

   CWDistance Complete  CompleteGroup  Score  
0          79        Y            1.0      7  
1          70        Y            1.0      8  
2          85        Y            1.0      7  
3          87        Y            1.0     10  
4          72        N            0.0      4

In [ ]:
# Drop several "unused" columns
#vars = ["ID", "GenderGroup", "GlassesGroup", "CompleteGroup"]
#df.drop(vars, axis=1, inplace = True)

# Add and eliminate rows

In some cases it is requiered to add new observations (rows) to the data set

In [ ]:
df.tail()

ID   Age Gender  GenderGroup Glasses  GlassesGroup  Height  Wingspan  \
47  48  24.0      M            2       N             0    79.5      75.0   
48  49  28.0      M            2       N             0    77.8      76.0   
49  50  30.0      F            1       N             0    74.6       NaN   
50  51   NaN      M            2       N             0    71.0      70.0   
51  52  27.0      M            2       N             0     NaN      71.5   

    CWDistance Complete  CompleteGroup  Score  
47          82        N            0.0      8  
48          99        Y            1.0      9  
49          71        Y            1.0      9  
50         101        Y            NaN      8  
51         103        Y            1.0     10

In [ ]:
df.loc[len(df.index)] = [26, 24, 'F', 1, 'Y', 1, 66, 'NaN', 68, 'N', 0, 3]

df.tail()


ID   Age Gender  GenderGroup Glasses  GlassesGroup  Height Wingspan  \
48  49  28.0      M            2       N             0    77.8     76.0   
49  50  30.0      F            1       N             0    74.6      NaN   
50  51   NaN      M            2       N             0    71.0     70.0   
51  52  27.0      M            2       N             0     NaN     71.5   
52  26  24.0      F            1       Y             1    66.0      NaN   

    CWDistance Complete  CompleteGroup  Score  
48          99        Y            1.0      9  
49          71        Y            1.0      9  
50         101        Y            NaN      8  
51         103        Y            1.0     10  
52          68        N            0.0      3

In [ ]:
# Eliminate inserted row
df.drop([28], inplace = True )

df.tail()


ID   Age Gender  GenderGroup Glasses  GlassesGroup  Height Wingspan  \
48  49  28.0      M            2       N             0    77.8     76.0   
49  50  30.0      F            1       N             0    74.6      NaN   
50  51   NaN      M            2       N             0    71.0     70.0   
51  52  27.0      M            2       N             0     NaN     71.5   
52  26  24.0      F            1       Y             1    66.0      NaN   

    CWDistance Complete  CompleteGroup  Score  
48          99        Y            1.0      9  
49          71        Y            1.0      9  
50         101        Y            NaN      8  
51         103        Y            1.0     10  
52          68        N            0.0      3

# Cleaning your data: drop out unused columns and/or drop out rows with any missing values

In [ ]:
# Drop unused columns
#vars = ["ID", "GenderGroup", "GlassesGroup", "CompleteGroup"]
#df.drop(vars, axis=1, inplace = True)

#vars = ["Age", "Gender", "Glasses", "Height", "Wingspan", "CWDistance", "Complete", "Score"]
#df = df[vars]

# Drop rows with any missing values
#df = df.dropna()

# Drop unused columns and drop rows with any missing values
#vars = ["Age", "Gender", "Glasses", "Height", "Wingspan", "CWDistance", "Complete", "Score"]
#df = df[vars].dropna()

df


ID   Age Gender  GenderGroup Glasses  GlassesGroup  Height Wingspan  \
0    1  56.0      F            1       Y             1   62.00     61.0   
1    2  26.0      F            1       Y             1   62.00     60.0   
2    3  33.0      F            1       Y             1   66.00     64.0   
3    4  39.0      F            1       N             0   64.00     63.0   
4    5  27.0      M            2       N             0   73.00     75.0   
5    6  24.0      M            2       N             0   75.00     71.0   
6    7  28.0      M            2       N             0   75.00     76.0   
7    8  22.0      F            1       N             0   65.00     62.0   
8    9  29.0      M            2       Y             1   74.00     73.0   
9   10  33.0      F            1       Y             1   63.00     60.0   
10  11  30.0      M            2       Y             1   69.50     66.0   
11  12  28.0      F            1       Y             1   62.75     58.0   
12  13  25.0      F            1       Y             1   65.00     64.5   
13  14  23.0      F            1       N             0   61.50     57.5   
14  15  31.0      M            2       Y             1   73.00     74.0   
15  16  26.0      M            2       Y             1   71.00     72.0   
16  17  26.0      F            1       N             0   61.50     59.5   
17  18  27.0      M            2       N             0   66.00     66.0   
18  19  23.0      M            2       Y             1   70.00     69.0   
19  20  24.0      F            1       Y             1   68.00     66.0   
20  21  23.0      M            2       Y             1   69.00     67.0   
21  22  29.0      M            2       N             0   71.00     70.0   
22  23  25.0      M            2       N             0   70.00     68.0   
23  24  26.0      M            2       N             0   69.00     71.0   
24  25  23.0      F            1       Y             1   65.00     63.0   
25  26  28.0      M            2       N             0   75.00     76.0   
26  27  24.0      M            2       N             0   78.40     71.0   
27  28  25.0      M            2       Y             1   76.00     73.0   
29  30  38.0      F            1       Y             1   61.50     61.0   
30  31  27.0      F            1       Y             1   62.00     60.0   
31  32  33.0      F            1       Y             1   65.30     64.0   
32  33  38.0      F            1       N             0   64.00     63.0   
33  34  27.0      M            2       N             0   77.00     75.0   
34  35  24.0      F            1       N             0   67.80     62.0   
35  36  27.0      M            2       N             0   68.00     66.0   
36  37  25.0      F            1       Y             1   65.00     64.5   
37  38  26.0      F            1       N             0   61.50     59.5   
38  39  31.0      M            2       Y             1   73.00     74.0   
39  40  30.0      M            2       Y             1   69.50     66.0   
40  41  23.0      F            1       N             0   70.40     71.0   
41  42  26.0      M            2       Y             1   73.50     72.0   
42  43  28.0      F            1       Y             1   72.50     72.0   
43  44  26.0      F            1       Y             1   72.00     72.0   
44  45  30.0      F            1       Y             1   66.00     64.0   
45  46  39.0      F            1       N             0   64.00     63.0   
46  47  27.0      M            2       N             0   78.00     75.0   
47  48  24.0      M            2       N             0   79.50     75.0   
48  49  28.0      M            2       N             0   77.80     76.0   
49  50  30.0      F            1       N             0   74.60      NaN   
50  51   NaN      M            2       N             0   71.00     70.0   
51  52  27.0      M            2       N             0     NaN     71.5   
52  26  24.0      F            1       Y             1   66.00      NaN   

    CWDistance Complete  Com

In [69]:
df.describe()

SepalLength  SepalWidth  PetalLength  PetalWidth
count   150.000000  150.000000   150.000000  150.000000
mean      5.843333    3.057333     3.758000    1.199333
std       0.828066    0.435866     1.765298    0.762238
min       4.300000    2.000000     1.000000    0.100000
25%       5.100000    2.800000     1.600000    0.300000
50%       5.800000    3.000000     4.350000    1.300000
75%       6.400000    3.300000     5.100000    1.800000
max       7.900000    4.400000     6.900000    2.500000

In [70]:
df.median()

<ipython-input-70-6d467abf240d>:1: FutureWarning: The default value of numeric_only in DataFrame.median is deprecated. In a future version, it will default to False. In addition, specifying 'numeric_only=None' is deprecated. Select only valid columns or specify the value of numeric_only to silence this warning.
  df.median()


SepalLength    5.80
SepalWidth     3.00
PetalLength    4.35
PetalWidth     1.30
dtype: float64

# Final remarks


- The understanding of your dataset is essential
    - Number of observations
    - Variables
    - Data types: numerical or categorial
    - What are my variables of interest

- There are several ways to do the same thing

- Cleaning your dataset (dropping out rows with any missing values) is a good practice

- The **Pandas** library provides fancy, high-performance, easy-to-use data structures and data analysis tools


# Activity: work with the iris dataset

Repeat this tutorial with the iris data set and respond to the following inquiries

1. Calculate the statistical summary for each quantitative variables. Explain the results
    - Identify the name of each column
      <br>SepalLength
      <br>SepalWidth
      <br>PetalLength
      <br>PetalWidth
      <br>Flower
    - Identify the type of each column
      <br>SepalLength: float
      <br>SepalWidth: float
      <br>PetalLength: float
      <br>PetalWidth: float
      <br>Flower: object
    - Minimum, maximum, mean, average, median, standar deviation
      - SepalLength
        <br> Minimum: 4.300000
        <br> Maximum: 7.900000
        <br> Mean: 5.843333
        <br> Average: 5.843333
        <br> Median: 5.80
        <br> Standard Deviation: 0.828066
      - SepalWidth
        <br> Minimum: 2.000000
        <br> Maximum: 4.400000
        <br> Mean: 3.057333
        <br> Average: 3.057333
        <br> Median: 3.00
        <br> Standard Deviation: 0.435866
      - PetalLength
        <br> Minimum: 1.000000
        <br> Maximum: 6.900000
        <br> Mean: 3.758000
        <br> Average: 3.758000
        <br> Median: 4.35
        <br> Standard Deviation: 1.765298
      - PetalWidth
        <br> Minimum: 0.100000
        <br> Maximum: 2.500000
        <br> Mean: 1.199333
        <br> Average: 1.199333
        <br> Median: 1.30
        <br> Standard Deviation: 0.762238
    
    
2. Are there missing data? If so, create a new dataset containing only the rows with the non-missing data


3. Create a new dataset containing only the petal width and length and the type of Flower


4. Create a new dataset containing only the setal width and length and the type of Flower


5. Create a new dataset containing the setal width and length and the type of Flower encoded as a categorical numerical column


# 2. New dataframe containing only the rows with non-missing data

In [86]:
# New dataframe with non-missing data

newdf = df.dropna()
newdf

SepalLength  SepalWidth  PetalLength  PetalWidth          Flower
0            5.1         3.5          1.4         0.2     Iris-setosa
1            4.9         3.0          1.4         0.2     Iris-setosa
2            4.7         3.2          1.3         0.2     Iris-setosa
3            4.6         3.1          1.5         0.2     Iris-setosa
4            5.0         3.6          1.4         0.2     Iris-setosa
..           ...         ...          ...         ...             ...
145          6.7         3.0          5.2         2.3  Iris-virginica
146          6.3         2.5          5.0         1.9  Iris-virginica
147          6.5         3.0          5.2         2.0  Iris-virginica
148          6.2         3.4          5.4         2.3  Iris-virginica
149          5.9         3.0          5.1         1.8  Iris-virginica

[150 rows x 5 columns]

# 3. New dataframe containing only the petal width and length and the type of Flower

In [85]:
dropCols = ['SepalLength', 'SepalWidth']
newdf = df.drop(dropCols, axis=1)
newdf

PetalLength  PetalWidth          Flower
0            1.4         0.2     Iris-setosa
1            1.4         0.2     Iris-setosa
2            1.3         0.2     Iris-setosa
3            1.5         0.2     Iris-setosa
4            1.4         0.2     Iris-setosa
..           ...         ...             ...
145          5.2         2.3  Iris-virginica
146          5.0         1.9  Iris-virginica
147          5.2         2.0  Iris-virginica
148          5.4         2.3  Iris-virginica
149          5.1         1.8  Iris-virginica

[150 rows x 3 columns]

# 4. New dataframe containing only the sepal width and length and the type of Flower

In [84]:
dropCols = ['PetalLength', 'PetalWidth']
newdf = df.drop(dropCols, axis=1)
newdf

SepalLength  SepalWidth          Flower
0            5.1         3.5     Iris-setosa
1            4.9         3.0     Iris-setosa
2            4.7         3.2     Iris-setosa
3            4.6         3.1     Iris-setosa
4            5.0         3.6     Iris-setosa
..           ...         ...             ...
145          6.7         3.0  Iris-virginica
146          6.3         2.5  Iris-virginica
147          6.5         3.0  Iris-virginica
148          6.2         3.4  Iris-virginica
149          5.9         3.0  Iris-virginica

[150 rows x 3 columns]

# 5. New dataset containing the sepal width and length and the type of Flower encoded as a categorical numerical column

In [87]:
dropCols = ['PetalLength', 'PetalWidth']
newdf = df.drop(dropCols, axis=1)
newdf['Flower'] = newdf['Flower'].astype('category')
newdf['Flower'] = newdf['Flower'].cat.codes
newdf

SepalLength  SepalWidth  Flower
0            5.1         3.5       0
1            4.9         3.0       0
2            4.7         3.2       0
3            4.6         3.1       0
4            5.0         3.6       0
..           ...         ...     ...
145          6.7         3.0       2
146          6.3         2.5       2
147          6.5         3.0       2
148          6.2         3.4       2
149          5.9         3.0       2

[150 rows x 3 columns]